platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/sources/CustomGeometrySource.java

```diff
@@ -0,0 +1,176 @@
+package com.mapbox.mapboxsdk.style.sources;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.UiThread;
+import android.support.annotation.WorkerThread;
+
+import com.mapbox.mapboxsdk.geometry.LatLngBounds;
+import com.mapbox.mapboxsdk.style.layers.Filter;
+import com.mapbox.services.commons.geojson.Feature;
+import com.mapbox.services.commons.geojson.FeatureCollection;
+
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Custom Vector Source, allows using FeatureCollections.
+ *
+ */
+@UiThread
+public class CustomGeometrySource extends Source {
+  private ThreadPoolExecutor executor;
+  private GeometryTileProvider provider;
+  private Map<TileID, AtomicBoolean> cancelledTileRequests;
+
+  /**
+   * Create a CustomGeometrySource
+   *
+   * @param id the source id
+   */
+  public CustomGeometrySource(String id, GeometryTileProvider provider_) {
+    executor = new ThreadPoolExecutor(2, 2, 60, TimeUnit.SECONDS, new ArrayBlockingQueue(80));
+    cancelledTileRequests = new HashMap<>();
```

Comment:
Is using a HashMap safe here.  Seems like it is getting modified by multiple threads, unless there is always only one WorkerThread.  setTileData is not marked as a WorkerThread.

Since the class is holding an AtomicBoolean it looks like it is intended to be used by multiple threads.

Reply:
Thanks for catching this @mkrussel77. `HashMap` will not work here and will need to use a thread-safe option like `ConcurrentHashMap`